content/blog/2016-05-09-when-to-use-varbit-chunks.md

```diff
@@ -0,0 +1,159 @@
+---
+title: When (not) to use varbit chunks
+created_at: 2016-05-09
+kind: article
+author_name: Björn “Beorn” Rabenstein
+---
+
+The embedded time-series database (TSDB) of the Prometheus server organizes the
+raw sample data of each time series in chunks of constant 1024 bytes size. In
+addition to the raw sample data, a chunk contains some meta-data, which allows
+the selection of a different encoding for each chunk. The most fundamental
+distinction is the encoding version. You select the version for newly
+created chunks via the command line flag
+`-storage.local.chunk-encoding-version`. So far, there were only two supported
+versions: 0 for the original delta encoding, and 1 for the improved
+double-delta encoding. With release
+[0.18.0](https://github.com/prometheus/prometheus/releases/tag/0.18.0), we
+added version 2, which is another variety of double-delta encoding. We call it
+_varbit encoding_ because it involves a variable bit-width per sample within
+the chunk. While version 1 is superior to version 0 in almost every aspect,
+there is a real trade-off between version 1 and 2. This blog post will help you
+to make that decision. Version 1 remains the default encoding, so if you want
+to try out version 2 after reading this article, you have to select it
+explicitly via the command line flag. There is no harm in switching back and
+forth, but note that chunks will not change their encoding version once they
+have been created.
+
+<!-- more -->
+
+## What is varbit encoding?
+
+From the beginning, we designed the chunked sample storage for easy addition of
+new encodings. When Facebook published a
+[paper on their in-memory TSDB Gorilla](http://www.vldb.org/pvldb/vol8/p1816-teller.pdf),
+we were intrigued by a number of similarities between the independently
+developed approaches of Gorilla and Prometheus. However, there were also many
+fundamental differences, which we studied in detail, wondering if we could get
+some inspiration from Gorilla to improve Prometheus.
+
+On the rare occasion of a free weekend ahead of me, I decided to give it a
+try. In a coding spree, I implemented what would later (after a considerable
+amount of testing and debugging) become the varbit encoding.
+
+In a future blog post, I will describe the technical details of the
+encoding. For now, you only need to know a few characteristics for your
+decision between the new varbit encoding and the traditional double-delta
+encoding. (I will call the latter just “double-delta encoding” from now on but
+note that the varbit encoding also uses double-deltas, just in a different
+way.)
+
+## What are the advantages of varbit encoding?
+
+In short: It offers a way better compression ratio. While the double-delta
+encoding needs about 3.3 bytes per sample for real-life data sets, the varbit
```

Comment:
That depends heavily on the data set (and on other circumstances like series churn - whenever you end a series, you leave behind a half-filled chunk, which lowers the average).
The reference server I used for my tests had 3.3 with double-delta and 1.28 with varbit.
I'll explain those details in the upcoming more technical blog post.